src/fn/sot457.ts

```diff
@@ -0,0 +1,231 @@
+      parameters: parameters,
+    }
+  }
+
+  const parameters = sot457_def.parse(raw_params)
+  return {
+    circuitJson: sot457WithoutParsing(parameters),
+    parameters: parameters,
+  }
+}
+export const getCcwSot457Coords = (parameters: {
+  p: number
+  w: number
+  pn: number
+}) => {
+  const { p, w, pn } = parameters
+
+  if (pn === 1) return { x: -w / 2 - 0.1, y: p }
+  if (pn === 2) return { x: -w / 2 - 0.1, y: 0 }
+  if (pn === 3) return { x: -w / 2 - 0.1, y: -p }
+  if (pn === 4) return { x: w / 2 + 0.1, y: -p }
+  if (pn === 5) return { x: w / 2 + 0.1, y: 0 }
+  if (pn === 6) return { x: w / 2 + 0.1, y: p }
+
+  throw new Error("Invalid pin number")
+}
+
+export const sot457WithoutParsing = (
+  parameters: z.infer<typeof sot457_def>,
+): AnyCircuitElement[] => {
+  const pads: AnyCircuitElement[] = []
+
+  if (parameters.wave) {
+    const pinPositions = {
+      1: ({ pw, ph }: { pw: number; ph: number }) =>
+        rectpad(
+          1,
+          -Number.parseFloat(parameters.p),
+          Number.parseFloat(parameters.p),
+          ph,
+          pw,
+        ),
+      2: ({ pw, ph }: { pw: number; ph: number }) =>
+        rectpad(
+          2,
+          -Number.parseFloat(parameters.p),
+          -Number.parseFloat(parameters.p),
+          ph,
+          pw,
+        ),
+      3: ({ pw, ph }: { pw: number; ph: number }) =>
+        pillpad(
+          3,
+          -Number.parseFloat(parameters.p),
+          0,
+          Number.parseFloat(parameters.pillw),
+          Number.parseFloat(parameters.pillh),
+        ),
+      4: ({ pw, ph }: { pw: number; ph: number }) =>
+        pillpad(
+          4,
+          Number.parseFloat(parameters.p),
+          0,
+          Number.parseFloat(parameters.pillw),
+          Number.parseFloat(parameters.pillh),
+        ),
+      5: ({ pw, ph }: { pw: number; ph: number }) =>
+        rectpad(
+          5,
+          Number.parseFloat(parameters.p),
+          Number.parseFloat(parameters.p),
+          ph,
+          pw,
+        ),
+      6: ({ pw, ph }: { pw: number; ph: number }) =>
+        rectpad(
+          6,
+          Number.parseFloat(parameters.p),
+          -Number.parseFloat(parameters.p),
+          ph,
+          pw,
+        ),
+    }
+
+    for (let i = 1; i <= parameters.num_pins; i++) {
+      if (pinPositions[i]) {
+        pads.push(
+          pinPositions[i]({
+            pw: Number.parseFloat(parameters.pl),
+            ph: Number.parseFloat(parameters.pw),
+          }),
+        )
+      }
+    }
+  } else {
+    for (let i = 1; i <= parameters.num_pins; i++) {
+      const { x, y } = getCcwSot457Coords({
+        p: Number.parseFloat(parameters.p),
+        w: Number.parseFloat(parameters.w),
+        pn: i,
+      })
+      pads.push(
+        rectpad(
+          i,
+          x,
+          y,
+          Number.parseFloat(parameters.pl),
+          Number.parseFloat(parameters.pw),
+        ),
+      )
+    }
+  }
```

Comment:
im sorry this is still a really bad function. Make this way way more readable. Use DRY principles.

Reply:
there's so much repeated parsing etc.